OpenRA.Mods.Common/Traits/CashTrickler.cs

```diff
@@ -9,58 +9,78 @@
  */
 #endregion
 
+using System;
 using OpenRA.Mods.Common.Effects;
 using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Traits
 {
 	[Desc("Lets the actor generate cash in a set periodic time.")]
-	class CashTricklerInfo : ITraitInfo
+	class CashTricklerInfo : UpgradableTraitInfo
 	{
 		[Desc("Number of ticks to wait between giving money.")]
-		public readonly int Period = 50;
+		public readonly int Interval = 50;
+
 		[Desc("Amount of money to give each time.")]
 		public readonly int Amount = 15;
-		[Desc("Whether to show the cash tick indicators (+$15 rising from actor).")]
+
+		[Desc("Whether to show the cash tick indicators rising from the actor.")]
 		public readonly bool ShowTicks = true;
-		[Desc("Amount of money awarded for capturing the actor.")]
-		public readonly int CaptureAmount = 0;
 
-		public object Create(ActorInitializer init) { return new CashTrickler(this); }
+		[Desc("How long to show the cash tick indicator when enabled.")]
+		public readonly int DisplayDuration = 30;
+
+		public override object Create(ActorInitializer init) { return new CashTrickler(this); }
 	}
 
-	class CashTrickler : ITick, ISync, INotifyCapture
+	class CashTrickler : UpgradableTrait<CashTricklerInfo>, ITick, ISync
 	{
 		readonly CashTricklerInfo info;
 		[Sync] int ticks;
+
 		public CashTrickler(CashTricklerInfo info)
+			: base(info)
 		{
 			this.info = info;
 		}
 
-		public void Tick(Actor self)
+		void ITick.Tick(Actor self)
 		{
+			if (IsTraitDisabled)
+				return;
+
 			if (--ticks < 0)
 			{
-				ticks = info.Period;
-				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(info.Amount);
-				MaybeAddCashTick(self, info.Amount);
+				ticks = info.Interval;
+				ModifyCash(self, self.Owner, info.Amount);
 			}
 		}
 
-		public void OnCapture(Actor self, Actor captor, Player oldOwner, Player newOwner)
+		void AddCashTick(Actor self, int amount)
 		{
-			if (info.CaptureAmount > 0)
-			{
-				newOwner.PlayerActor.Trait<PlayerResources>().GiveCash(info.CaptureAmount);
-				MaybeAddCashTick(self, info.CaptureAmount);
-			}
+			self.World.AddFrameEndTask(w => w.Add(
+				new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), info.DisplayDuration)));
 		}
 
-		void MaybeAddCashTick(Actor self, int amount)
+		void ModifyCash(Actor self, Player newOwner, int amount)
 		{
-			if (info.ShowTicks)
-				self.World.AddFrameEndTask(w => w.Add(new FloatingText(self.CenterPosition, self.Owner.Color.RGB, FloatingText.FormatCashTick(amount), 30)));
+			var resources = newOwner.PlayerActor.Trait<PlayerResources>();
```

Comment:
Can we cache this instead of looking it up each cycle?  You can refer to the `Refinery` trait to see an example of how to set and change it on ownership change.